controllers/trainingjob/trainingjob_controller.go

```diff
@@ -18,411 +18,347 @@ package trainingjob
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
+		if !controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
+			if err = r.handleUpdates(ctx); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to update SageMaker trainingjob"))
+			}
+		} else {
+			if _, err := ctx.SageMakerClient.StopTrainingJob(ctx, ctx.TrainingJobName); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to delete training job"))
+			}
 		}
-		return RequeueAfterInterval(r.PollInterval, err)
 
-	case sagemaker.TrainingJobStatusStopped, sagemaker.TrainingJobStatusFailed:
-		return NoRequeue()
+		break
 
 	case sagemaker.TrainingJobStatusCompleted:
-		// If job has completed populate the model full path
-		log.Info("Training has completed updating model path")
-		// SageMaker stores the model artifact in OutputDataConfig path with path /output/model.tar.gz
-		// SageMaker documentation https://docs.aws.amazon.com/sagemaker/latest/dg/cdf-training.html
-		const outputPath string = "/output/model.tar.gz"
-		modelPath := *state.Spec.OutputDataConfig.S3OutputPath + state.Status.SageMakerTrainingJobName + outputPath
-		if err = r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-			ModelPath:                modelPath,
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
-			return RequeueIfError(err)
+		if err = r.updateCompleted(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to update training job"))
 		}
-		return NoRequeue()
+
+		fallthrough
+
+	case sagemaker.TrainingJobStatusStopped, sagemaker.TrainingJobStatusFailed:
+		if controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
+			return r.cleanupAndRemoveFinalizer(ctx)
+		}
+
+		break
 
 	default:
-		unknownStateError := errors.New(string("Unknown Training Job Status " + trainingJobDescription.TrainingJobStatus))
-		log.Error(unknownStateError, "Job is in unknown status")
-		return NoRequeue()
+		unknownStateError := errors.New(string("Unknown Training Job Status " + ctx.TrainingJobDescription.TrainingJobStatus))
+		return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(unknownStateError, "Job is in an unknown status"))
```

Comment:
This wrapped error will say something like `Job is an in unknown status: Unknown Training Job Status <status>`. We should remove one of the messages to be concise.